File: src/lib/brain/vector-search.ts
/**
 * @fileOverview Production Retrieval Wrappers
 * Canonical wrappers over Vertex AI Vector Search + Firestore hydration.
 */

import { vertexSearch, VectorSearchResult } from './vertex-search-service';

/**
 * Tier 1: Curated Articles
 * Highest-trust retrieval source.
 */
export async function searchArticles(args: {
  query: string;
  hubId: string;
  spaceId: string;
  allowedHelpCenterIds?: string[];
  limit?: number;
}): Promise<Array<{
  id: string;
  text: string;
  title?: string;
  sourceType: 'article';
  libraryId?: string;
  visibility: 'public' | 'private';
  url?: string;
  score: number;
}>> {
  const {
    query,
    hubId,
    spaceId,
    allowedHelpCenterIds,
    limit = 8,
  } = args;

  // If article access is explicitly restricted to no libraries, return nothing.
  if (Array.isArray(allowedHelpCenterIds) && allowedHelpCenterIds.length === 0) {
    return [];
  }

  const results: VectorSearchResult[] = await vertexSearch.search({
    query,
    sourceType: 'article',
    spaceId,
    hubId,
    libraryIds: allowedHelpCenterIds,
    limit,
  });

  return results.map((r) => ({
    id: r.id,
    text:
      r.metadata.content ||
      r.metadata.body ||
      r.metadata.summary ||
      r.metadata.subtitle ||
      r.metadata.title ||
      '',
    title: r.metadata.title,
    sourceType: 'article' as const,
    libraryId: r.metadata.libraryId || r.metadata.destinationLibraryId || r.metadata.helpCenterId,
    visibility: (r.metadata.visibility || 'public') as 'public' | 'private',
    url:
      r.metadata.url ||
      r.metadata.slug ||
      (r.metadata.helpCenterId ? `/hc/${r.metadata.helpCenterId}/articles/${r.id}` : undefined),
    score: r.score,
  }));
}

/**
 * Tier 2: Topics
 * Recurring grouped patterns derived from Insights.
 */
export async function searchTopics(args: {
  query: string;
  spaceId: string;
  hubId?: string | null;
  limit?: number;
}): Promise<Array<{
  id: string;
  text: string;
  title?: string;
  sourceType: 'topic';
  signalLevel?: 'low' | 'medium' | 'high';
  insightCount?: number;
  score: number;
}>> {
  const {
    query,
    spaceId,
    hubId,
    limit = 5,
  } = args;

  const results: VectorSearchResult[] = await vertexSearch.search({
    query,
    sourceType: 'topic',
    spaceId,
    hubId,
    limit,
  });

  return results.map((r) => ({
    id: r.id,
    title: r.metadata.title,
    text: r.metadata.summary || r.metadata.title || '',
    sourceType: 'topic' as const,
    signalLevel: r.metadata.signalLevel as 'low' | 'medium' | 'high' | undefined,
    insightCount: typeof r.metadata.insightCount === 'number' ? r.metadata.insightCount : undefined,
    score: r.score,
  }));
}

/**
 * Tier 3: Insights
 * Internal support learnings. These are lower-trust than Articles and Topics.
 */
export async function searchInsights(args: {
  query: string;
  hubId: string;
  spaceId: string;
  limit?: number;
}): Promise<Array<{
  id: string;
  text: string;
  title?: string;
  sourceType: 'insight';
  visibility: 'private' | 'public';
  signalLevel?: 'low' | 'medium' | 'high';
  origin?: 'automatic' | 'manual' | 'imported';
  score: number;
}>> {
  const {
    query,
    hubId,
    spaceId,
    limit = 5,
  } = args;

  const results: VectorSearchResult[] = await vertexSearch.search({
    query,
    sourceType: 'insight',
    spaceId,
    hubId,
    visibility: 'private',
    limit,
  });

  return results.map((r) => ({
    id: r.id,
    title: r.metadata.title,
    text: r.metadata.content || r.metadata.summary || r.metadata.title || '',
    sourceType: 'insight' as const,
    visibility: (r.metadata.visibility || 'private') as 'private' | 'public',
    signalLevel: r.metadata.signalLevel as 'low' | 'medium' | 'high' | undefined,
    origin: r.metadata.origin as 'automatic' | 'manual' | 'imported' | undefined,
    score: r.score,
  }));
}

File: src/lib/brain/vector_search.py
"""
Vector Search Wrappers - Production retrieval wrappers over Vertex AI Vector Search.
"""

from typing import List, Optional, Literal
from .vertex_search_service import vertex_search, SearchParams, VectorSearchResult


class ArticleResult:
    """Search result for an article."""
    def __init__(self, id: str, text: str, title: Optional[str], score: float, 
                 libraryId: Optional[str], visibility: Literal['public', 'private'], 
                 url: Optional[str]):
        self.id = id
        self.text = text
        self.title = title
        self.sourceType = 'article'
        self.libraryId = libraryId
        self.visibility = visibility
        self.url = url
        self.score = score


class TopicResult:
    """Search result for a topic."""
    def __init__(self, id: str, text: str, title: Optional[str], score: float,
                 signalLevel: Optional[Literal['low', 'medium', 'high']], 
                 insightCount: Optional[int]):
        self.id = id
        self.title = title
        self.text = text
        self.sourceType = 'topic'
        self.signalLevel = signalLevel
        self.insightCount = insightCount
        self.score = score


class InsightResult:
    """Search result for an insight."""
    def __init__(self, id: str, text: str, title: Optional[str], score: float,
                 visibility: Literal['private', 'public'],
                 signalLevel: Optional[Literal['low', 'medium', 'high']],
                 origin: Optional[Literal['automatic', 'manual', 'imported']]):
        self.id = id
        self.title = title
        self.text = text
        self.sourceType = 'insight'
        self.visibility = visibility
        self.signalLevel = signalLevel
        self.origin = origin
        self.score = score


async def search_articles(
    query: str,
    hubId: str,
    spaceId: str,
    allowedHelpCenterIds: Optional[List[str]] = None,
    limit: int = 8
) -> List[ArticleResult]:
    """
    Tier 1: Curated Articles
    Highest-trust retrieval source.
    
    Args:
        query: Search query
        hubId: Hub ID
        spaceId: Space ID
        allowedHelpCenterIds: Allowed library IDs (None = restrict access)
        limit: Max results
        
    Returns:
        List of ArticleResult objects
    """
    # If article access is explicitly restricted to no libraries, return nothing
    if isinstance(allowedHelpCenterIds, list) and len(allowedHelpCenterIds) == 0:
        return []
    
    results: List[VectorSearchResult] = await vertex_search.search(
        SearchParams(
            query=query,
            sourceType='article',
            spaceId=spaceId,
            hubId=hubId,
            libraryIds=allowedHelpCenterIds,
            limit=limit
        )
    )
    
    return [
        ArticleResult(
            id=r.id,
            text=r.metadata.get('body') or r.metadata.get('summary') or r.metadata.get('title') or '',
            title=r.metadata.get('title'),
            score=r.score,
            libraryId=r.metadata.get('libraryId') or r.metadata.get('destinationLibraryId'),
            visibility=r.metadata.get('visibility', 'public'),
            url=r.metadata.get('url') or r.metadata.get('slug')
        )
        for r in results
    ]


async def search_topics(
    query: str,
    spaceId: str,
    hubId: Optional[str] = None,
    limit: int = 5
) -> List[TopicResult]:
    """
    Tier 2: Topics
    Recurring grouped patterns derived from Insights.
    
    Args:
        query: Search query
        spaceId: Space ID
        hubId: Optional Hub ID
        limit: Max results
        
    Returns:
        List of TopicResult objects
    """
    results: List[VectorSearchResult] = await vertex_search.search(
        SearchParams(
            query=query,
            sourceType='topic',
            spaceId=spaceId,
            hubId=hubId,
            limit=limit
        )
    )
    
    return [
        TopicResult(
            id=r.id,
            title=r.metadata.get('title'),
            text=r.metadata.get('summary') or r.metadata.get('title') or '',
            score=r.score,
            signalLevel=r.metadata.get('signalLevel'),
            insightCount=r.metadata.get('insightCount')
        )
        for r in results
    ]


async def search_insights(
    query: str,
    hubId: str,
    spaceId: str,
    limit: int = 5
) -> List[InsightResult]:
    """
    Tier 3: Insights
    Internal support learnings. Lower-trust than Articles and Topics.
    
    Args:
        query: Search query
        hubId: Hub ID
        spaceId: Space ID
        limit: Max results
        
    Returns:
        List of InsightResult objects
    """
    results: List[VectorSearchResult] = await vertex_search.search(
        SearchParams(
            query=query,
            sourceType='insight',
            spaceId=spaceId,
            hubId=hubId,
            visibility='private',
            limit=limit
        )
    )
    
    return [
        InsightResult(
            id=r.id,
            title=r.metadata.get('title'),
            text=r.metadata.get('content') or r.metadata.get('summary') or r.metadata.get('title') or '',
            score=r.score,
            visibility=r.metadata.get('visibility', 'private'),
            signalLevel=r.metadata.get('signalLevel'),
            origin=r.metadata.get('origin')
        )
        for r in results
    ]
